ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -544,4 +531,127 @@ private boolean containsName(String name) {
+        accountMap = new HashMap<>();
+      }
+      try {
+        remoteAccountInfoMap = new AccountInfoMap(accountMap);
+      } catch (JSONException e) {
+        // Do not depend on Helix to log, so log the error message here.
+        logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+        accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+      }
+      // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+      // be caught by Helix and helixStore#update will return false.
+      if (hasConflictingAccount(updatedAccounts, remoteAccountInfoMap)) {
+        // Throw exception, so that helixStore can capture and terminate the update operation
+        throw new IllegalArgumentException(
+            "Updating accounts failed because one account to update conflicts with existing updatedAccounts");
+      } else {
+        for (Account account : updatedAccounts) {
+          try {
+            accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+          } catch (Exception e) {
+            String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
+                + account.getId() + " accountName=" + account.getName();
+            // Do not depend on Helix to log, so log the error message here.
+            logger.error(message, e);
+            throw new IllegalStateException(message, e);
+          }
+        }
+        newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+        return newRecord;
+      }
+    }
+
+    /**
+     * Save the state from zookeeper from before the update and after the update to disk.
+     * This will write the backup content string to the disk with the following file name format:
+     * {@code {yyyyMMdd}T{HHmmss}-{unique long}.bak}.
+     * If there are multiple files with the same timestamp, the unique long will prevent the file names from clashing.
+     * @param succeeded {@code true} iff the update succeeded.
+     */
+    void persistBackup(boolean succeeded) {
+      try {
+        String timestamp = LocalDateTime.now().format(TIMESTAMP_FORMATTER);
+        JSONObject backupContent = getBackupContent(succeeded);
+        try (BufferedWriter writer = getWriter(timestamp)) {
+          backupContent.write(writer);
+        }
+      } catch (JSONException | IOException e) {
+        logger.error("Could not write backup file", e);
+      }
+    }
+
+    /**
+     * @param timestamp the timestamp for the backup file
+     * @return a {@link BufferedWriter} for the backup file.
+     * @throws IOException if the file could not be created.
+     */
+    private BufferedWriter getWriter(String timestamp) throws IOException {
```

Comment:
This might be ok if the granularity of the timestamp is fine relative to the call rate of this method, but if it's going to be called many times a second at one second granularity seems like this can get slow.  I'm guessing it is slow since it's about making files?

Reply:
For our use case, we should not see more than a few account updates per hour and only a see more than one per second in very rare cases (like our unit tests). An alternative would be to use log4j to write all updates to one log, but I think having separate files per update is easier to read and debug with.